client/store/products/reducer.js

```diff
@@ -2,42 +2,30 @@
 /**
  * External dependencies
  */
-import { get } from 'lodash';
+import { merge } from 'lodash';
 
 /**
  * Internal dependencies
  */
 import { ERROR } from 'store/constants';
 import { getJsonString } from 'store/utils';
 
-export const DEFAULT_STATE = {
-	queries: {},
-};
+export const DEFAULT_STATE = {};
 
 export default function productsReducer( state = DEFAULT_STATE, action ) {
-	if ( 'SET_PRODUCTS' === action.type ) {
-		const prevQueries = get( state, 'queries', {} );
-		const queryKey = getJsonString( action.query );
-		const queries = {
-			...prevQueries,
-			[ queryKey ]: [ ...action.products ],
-		};
-		return {
-			...state,
-			queries,
-		};
-	}
-	if ( 'SET_PRODUCTS_ERROR' === action.type ) {
-		const prevQueries = get( state, 'queries', {} );
-		const queryKey = getJsonString( action.query );
-		const queries = {
-			...prevQueries,
-			[ queryKey ]: ERROR,
-		};
-		return {
-			...state,
-			queries,
-		};
+	const queryKey = getJsonString( action.query );
+
+	switch ( action.type ) {
+		case 'SET_PRODUCTS':
+			return merge( {}, state, {
```

Comment:
Right, actually since state must be immutable, we are writing all values from `state` to the first parameter empty object `{}`, except the property that we are updating with the third parameter `{ [ queryKey ]: action.products }`.